File: src/rc/controls.py
import math
import random
from sys import maxsize
from typing import List

from .rkt_types import Control

# Controls (piecewise constant, all times are in seconds)
uptime = 10.0
numturns1 = 10
turntime1 = lambda bt: (bt - uptime) / numturns1
turntime1a = turntime1
turntime1b = lambda bt: bt / numturns1
numturns2 = 10
turntime2 = 5.0

# Simplest possible control: straight up, then straight sideways.
up = [Control(0.0, uptime, 0.0)]
over = [Control(uptime, 1000.0, math.pi / 2.0)]
controls_up_over = up + over

def piecewise_turn(init_time: float, init_angle: float, final_angle: float,
                   numturns: int, turntime: float) -> List[Control]:
    return [Control(init_time + r*turntime, init_time + (r+1.0)*turntime,
                    init_angle + (final_angle - init_angle)*(r+1.0)/numturns)
            for r in range(numturns)]

gravity_turn1 = lambda bt: piecewise_turn(0.0, 0.0, math.pi/4.0, numturns1, turntime1(bt))
gravity_turn1a = lambda bt: piecewise_turn(0.0, 0.0, math.pi/6.0, numturns1, turntime1a(bt))
gravity_turn1b = lambda bt: piecewise_turn(0.0, math.pi/6.0, math.pi/4.0, numturns1, turntime1b(bt))

staging_time = 3.0
staging_delay = [Control(0.0, staging_time, 0.0, 0.0)]

gravity_turn2 = piecewise_turn(0.0, math.pi/4.0, math.pi/2.0, numturns2, turntime2)
over12 = [Control(0.0, 1000.0, math.pi / 2.0)]
boostback = [Control(0.0, 1000.0, - math.pi / 2.0, 0.4)] # minus = backward, and using 40% thrust

def reduce_times(controls: List[Control]) -> List[Control]:
    time = 0.0
    new_controls = []
    for c in controls:
        delta_t = c.t2 - c.t1
        control = Control(time, time + delta_t, c.force_phi, c.force_mag)
        new_controls.append(control)
        time += delta_t
    return new_controls

def perturb_controls_phi_random(controls: List[Control], scale: float, index_min: int = 0,
                                index_max: int = maxsize) -> List[Control]:
    new_controls = []
    for i, c in enumerate(controls):
        # Skip controls out of bounds
        if not index_min <= i < index_max:
            new_control = Control(c.t1, c.t2, c.force_phi, c.force_mag)
        else:
            delta_phi = (random.random() * 2.0 - 1.0) * (math.pi / 2.0) / scale
            new_control = Control(c.t1, c.t2, c.force_phi + delta_phi, c.force_mag)
        new_controls.append(new_control)
    return new_controls

File: src/controls.rs
pub mod mcontrols {
    use std::f64::consts::PI;
    use rand::prelude::*;
    use crate::rkt_types::mrkt_types::{Control};

    // Controls (piecewise constant, all times are in seconds)
    pub const uptime: f64 = 10.0;
    pub const numturns1: i32 = 10;
    pub fn turntime1(bt: f64) -> f64 {
        (bt - uptime) / numturns1 as f64
    }
    pub fn turntime1a(bt: f64) -> f64 {
        turntime1(bt)
    }
    pub fn turntime1b(bt: f64) -> f64 {
        bt / numturns1 as f64
    }
    pub const numturns2: i32 = 10;
    pub const turntime2: f64 = 5.0;

    // Simplest possible control: straight up, then straight sideways.
    pub const up: Control = Control{t1: 0.0, t2: uptime, force_phi: 0.0, force_mag: 1.0};
    pub const over: Control = Control{t1: uptime, t2: 1000.0, force_phi: PI / 2.0, force_mag: 1.0};
    pub const controls_up_over: [Control; 2] = [up, over];

    pub fn piecewise_turn(init_time: f64, init_angle: f64, final_angle: f64, numturns: i32, turntime: f64) -> Vec<Control> {
        (0..numturns).into_iter().map(|r| r as f64).map(|r| Control {
            t1: init_time + r*turntime,
            t2: init_time + (r+1.0)*turntime,
            force_phi: init_angle + (final_angle - init_angle)*(r+1.0)/numturns as f64,
            force_mag: 1.0,
        }).collect()
    }

    pub fn gravity_turn1(bt: f64) -> Vec<Control> {
        piecewise_turn(0.0, 0.0, PI/4.0, numturns1, turntime1(bt))
    }
    pub fn gravity_turn1a(bt: f64) -> Vec<Control> {
        piecewise_turn(0.0, 0.0, PI/6.0, numturns1, turntime1a(bt))
    }
    pub fn gravity_turn1b(bt: f64) -> Vec<Control> {
        piecewise_turn(0.0, PI/6.0, PI/4.0, numturns1, turntime1b(bt))
    }

    pub const staging_time: f64 = 3.0;
    pub const staging_delay: Control = Control{t1: 0.0, t2: staging_time, force_phi: 0.0, force_mag: 0.0};

    pub fn gravity_turn2() -> Vec<Control> {
        piecewise_turn(0.0, PI/4.0, PI/2.0, numturns2, turntime2)
    }
    pub const over12: Control = Control{t1: 0.0, t2: 1000.0, force_phi: PI / 2.0, force_mag: 1.0};
    pub const boostback: Control = Control{t1: 0.0, t2: 1000.0, force_phi: - PI / 2.0, force_mag: 0.4}; // minus = backward, and using 40% thrust

    pub fn reduce_times(controls: & Vec<Control>) -> Vec<Control> {
        let mut time: f64 = 0.0;
        //let new_controls = &mut vec![];
        //for c in controls {
        let mkctrl = |x:(usize, &Control)| {
            let (i, c) = x;
            let delta_t = c.t2 - c.t1;
            let control = Control{t1: time, t2: time + delta_t, force_phi: c.force_phi, force_mag: c.force_mag};
            //new_controls.push(control);
            time += delta_t;
            control
        };
        controls.iter().enumerate().map(mkctrl).collect()
        //*new_controls
    }

    pub fn perturb_controls_phi_random(controls: & Vec<Control>, scale: f64, index_min: i32, index_max: i32) -> Vec<Control> {
        //let new_controls = &mut vec![];
        //for (i, c) in controls.iter().enumerate() {
        let mkctrl = |x:(usize, &Control)| {
            let (i, c) = x;
            // Skip controls out of bounds
            if ! (index_min as usize <= i && i < index_max as usize) {
                let new_control = Control{t1: c.t1, t2: c.t2, force_phi: c.force_phi, force_mag: c.force_mag};
                //new_controls.push(new_control);
                new_control
            } else {
                let mut rng = rand::thread_rng();
                let y: f64 = rng.gen(); // generates a float between 0 and 1
                let delta_phi = (y * 2.0 - 1.0) * (PI / 2.0) / scale;
                let new_control = Control{t1: c.t1, t2: c.t2, force_phi: c.force_phi + delta_phi, force_mag: c.force_mag};
                //new_controls.push(new_control);
                new_control
            }
        };
        controls.iter().enumerate().map(mkctrl).collect()
        //*new_controls
    }
}